### ISSUE TITLE
parallel numba implementation of MAP algo with weights (performs slightly worse than pyhdfe)

### ISSUE THREAD

**Issue by s3alfisc** [jeffgortmaker/pyhdfe#3]
Hi Jeff, 

I have implemented the plain MAP demeaning algorithm in numba. (Unfortunately) I do not get it to perform better than pyhdfe's implementation, but I have added support for weighting, which I need for non-linear models. As it is quite short, here is the implementation (you can find it in pyfixest [here](https://github.com/s3alfisc/pyfixest/blob/master/pyfixest/demean.py)): 

```py
import numpy as np
from numba import njit, prange

@njit(parallel = True, cache = False, fastmath = False)
def demean(cx, flist, weights, tol = 1e-08, maxiter = 2000):

    '''
    Demean a Matrix cx by fixed effects in flist.
    The fixed effects are weighted by weights. Convervence tolerance
    is set to 1e-08 for the sum of absolute differences.
    Args:
        cx: Matrix to be demeaned
        flist: Matrix of fixed effects
        weights: Weights for fixed effects
        tol: Convergence tolerance. 1e-08 by default.
    Returns
        res: Demeaned matrix of dimension cx.shape
    '''
    N = cx.shape[0]
    fixef_vars = flist.shape[1]
    K = cx.shape[1]

    res = np.zeros((N,K))

    # loop over all variables to demean, in parallel 
    for k in prange(K):

        cxk = cx[:,k].copy()
        oldxk = cxk - 1

        converged = False
        for _ in range(maxiter):
           
            for i in range(fixef_vars):
                fmat = flist[:,i]
                weighted_ave = _ave3(cxk, fmat, weights)
                cxk = cxk - weighted_ave

            if np.sum(np.abs(cxk - oldxk)) < tol:
                converged = True
                break

            # update 
            oldxk = cxk.copy()



        res[:,k] = cxk

    return res
```

`_ave3()` simply implements a group wise weighted average: 

```py
@njit
def _ave3(x, f, w):

    N = len(x)

    wx_dict = {}
    w_dict = {}

    # Compute weighted sums using a dictionary
    for i in range(N):
        j = f[i]
        if j in wx_dict:
            wx_dict[j] += w[i] * x[i]
        else:
            wx_dict[j] = w[i] * x[i]

        if j in w_dict:
            w_dict[j] += w[i]
        else:
            w_dict[j] = w[i]

    # Convert the dictionaries to arrays
    wx = np.zeros_like(f, dtype=x.dtype)
    w = np.zeros_like(f, dtype=w.dtype)

    for i in range(N):
        j = f[i]
        wx[i] = wx_dict[j]
        w[i] = w_dict[j]

    # Compute the average
    wxw_long = wx / w

    return wxw_long
```

Here is one benchmark against `pyhdfe`: 

```py
%load_ext autoreload
%autoreload 2

from pyfixest.demean import demean
import numpy as np
import time

np.random.seed(1238)
N = 1_000_000
x = np.random.normal(0, 1, 4*N).reshape((N,4))
f1 = np.random.choice(list(range(1000)), N).reshape((N,1))
f2 = np.random.choice(list(range(1000)), N).reshape((N,1))

flist = np.concatenate((f1, f2), axis = 1)
weights = np.ones(N)

import pyhdfe
start_time = time.time()
algorithm = pyhdfe.create(flist)
res_pyhdfe = algorithm.residualize(x)
end_time = time.time()
end_time - start_time
# 2.168398380279541


start_time = time.time()
algorithm = pyhdfe.create(flist)
res_pyfixest = demean(x, flist, weights, tol = 1e-08)
# Calculate the execution time
end_time = time.time()
end_time - start_time
# 3.359426736831665

np.allclose(res_pyhdfe , res_pyfixest)
# True
```

In case you have any idea how to speed this up, please let me know =) 

I'll try to optimize this a little more, but then would like to open a PR to `pyhdfe`. I thought that for a start, I would add a 'weights' argument to the 'residualize' method, which would then call the `demean()` function above and else keep everything as is. Would you be interested in that, and if yes, do you have any thoughts / comments? 

Best, Alex

**Comment by s3alfisc** [comment]
Just for comparison - `fixest` solves the same problem in around 0.1 seconds: 

```r
library(fixest)
library(reticulate)
library(pracma)
x <- py$x
flist <- py$flist
weights <- py$weights
#res <- py$res

tic()
res_fixest <- fixest::demean(X = x, f = flist, weights = as.vector(weights))
toc()
# elapsed time is 0.140000 seconds 

```

**Comment by jeffgortmaker** [comment]
I think there's three things going on here, each of which could potentially be a useful PR?

1. Support for weighting. 
1.a. Mind reminding me what specific use-cases there are for this? Do I remember correctly that the likely main use case is interacting fixed effects with continuous variables? 
1.b. If weights change with the matrix being residualized, they would indeed make sense as an optional `residualize` argument. Otherwise they might make more sense in `create`?
1.c. There should be basic support for weighting for existing algorithms before adding another package into the mix. Which existing algorithms will support it? At a minimum you'd want to modify the existing simple MAP algorithms to add support. I think should be a simple extension of the `Groups` class?
1.d. The other main thing to add would be unit tests. Figuring out the use-case from (1.a) that most closely resembles existing unit tests would be the way to go, I think.

2. Option to use numba when possible.
2.a. Before putting together a PR for this, I'd like to see an implementation that yields *significant* speedups over the status quo. So something like the fixest timing you cited. Otherwise I'm cautious to add code that depends on another package, since this adds maintenance costs for me, documentation complexity for users, and little potential speedup benefits.
2.b. Given a clear speedup, I think adding a `numba` flag to `create`'s `options` argument would make the most sense. It'd be by default `False`, and if `True` would check whether numba is installed, raising an informative error message if not. Numba would then be an optional requirement.
2.c. The place to start is probably implementing a numba clone of the existing MAP algorithm, using the same class structure as the rest of PyHDFE. Then you'd be comparing apples-to-apples in speedtests.
2.d. I'm not sure how to speed up the above code because I'm not very familiar with numba. Personally I go straight to jax when I need big speedups, but I wouldn't recommend that here. What I do recommend is making the simplest possible implementations of numpy-MAP, numba-MAP, and fixest-MAP, and profiling all of these to see exactly where speedups are coming from.

3. Option to use parallelization.
3.a. My sense is a `multiprocessing` approach here would add a lot of overhead, so parallelization through numba (i.e. through treading) probably makes the most sense.
3.b. Of course numpy also uses multithreading, so again I'd like to see an implementation that yields *significant* speedups over the status quo before you put together a PR, for the same reason as (2.a). This type of investigation could also generate some useful insights about how to better use numpy's multithreading capabilities!
3.c. Given a clear speedup, I think adding a `processes` flag to `create`'s `options` argument would make the most sense. It'd be by default `1`, and if configured higher, would require that the above `numba` flag is `True`.

What do you think? Hope this gives a good sense of how to proceed! I don't have much free time on my end to help write the actual PRs, but happy to accept ones that look like what I've outlined above, or something similar.

**Comment by s3alfisc** [comment]
Hi @jeffgortmaker - thanks for providing all this clarity! I fully agree that I am mixing up too many different things and that it makes sense to split this up into different pieces/PRs. 

The most obvious starting point would be 1). Allowing for weights in the demeaning algo will allow to implement WLS estimation with HDFEs. This can then in turn unlock non-linear glm estimation with HDFEs via iterated weighted least squares. This is what Stata's [pplmhdfe](https://arxiv.org/pdf/1903.01690.pdf) does, and I think `fixest` also implements a variant of an iterated WLS algorithm for its non-linear models. Do you have some reading recommendation for the fixed effect x continuous variable interaction that you mention? That's a problem I have so far successfully managed to circumvent... :D 

I just took a closer look at the `pyhdfe` source code, and you are right that adding a `wmean` method to the `Groups` class should bring me very, very far. The weights should be fixed in one single call of `residualize` (at least in the use case I envision), and I thought about adding a `weights` argument to the appropriate `residualize` methods and pass it to the respective algorithms. Then I'll write some unit tests + open the PR. 

On 2) and 3) - I will put the `numba` part on hold for now. I have tried quite a few things over the weekend and am not too optimistic that I'll manage to bring this up to `fixest` speed without serious support from a more experienced `numba` programmer. Though if I should ever get there, I will make sure to let you know (one is still allowed to dream, I guess😄).

**Comment by jeffgortmaker** [comment]
Agree that starting with (1) makes sense. Seems like [reghdfe](http://scorreia.com/help/reghdfe.html#absorb) supports FE x continuous interactions, so you can take a look at that code to see if the same type of weights are needed.

And sounds great. A `weights=None` default argument to `Algorithm.residualize` that are passed through to something like a `weights=None` default argument to `Groups.mean` is probably the easiest way to go?

Thanks for working on this!

**Comment by jeffgortmaker** [comment]
Closing after merging in #4

**Comment by s3alfisc** [comment]
Hi @jeffgortmaker , I just merged a PR by @styfenschaer into pyfixest, which speeds up the alternating projection algo by around 5x (or more). 

Would you be open to a PR to `PyHDFE`? This would require to add `numba` as a direct dependency. 

Here is a [benchmark](https://github.com/s3alfisc/pyfixest/releases/tag/v0.10), and [here](https://github.com/s3alfisc/pyfixest/blob/b202c342487b683467782967689e188d32db4088/pyfixest/demean.py#L178) is @styfenschaer's implementation. 

Best, Alex

**Comment by jeffgortmaker** [comment]
I'm always open to PRs! But I also have in mind https://github.com/s3alfisc/pyfixest/issues/98, where @lrberge mentioned that he has a clear idea of how to put together a pure C++ API of the algorithm he's using for fixest. If this and/or other pure C++ implementations will supersede a Numba implementation like this, I would prefer to wait to see how they compare.

I don't have any specific preferences for a lower-level (i.e. not just NumPy) framework to use, whether it's Numba, pure C++, or something else. (JAX and its relations aside, which I like but currently have sharp edges, my sense is that these two options are the most likely ones, but I'm happy to learn about others.) I do have the following high-level preferences, though:

1. Limit dependencies. This is a very simple package, so it should not depend on complicated packages.
2. Prefer easier-to-read, shorter, and more extensible code. For this reason, I may prefer Numba over pure C++.
3. If possible, wait on very clear benchmarks/tests for anything that stretches (1) or (2).

For now, I don't think there's any rush for speedups alone, and like @lrberge I'm quite busy, so let's wait on any PyHDFE PR until things have settled in the above thread. Until then, I would be interested to better understand your benchmark, as per (3). In particular:

- Are the two functions using the same convergence criteria/tolerance, algorithm, threads/processes, and memory usage?
- If so, where specifically are the speedups coming from? I'd be very happy to learn why we get large speedups on a sizable dataset where speed matters (e.g. when residualization takes more than a minute) and there aren't any differences other than pure NumPy vs. Numba.

**Comment by s3alfisc** [comment]
Sounds good, let's wait with the PR then! 

On your two questions:
- Actually, convergence criteria are not completely identical: while both implementations check absolute differences with tolerance 1e-08, the convergence criterion in the `numba` implementation is column-wise, while in `pyhdfe`, it is defined on the entire matrix. So I think that the `numba` algos criterion should be stricter?
- @styfenschaer probably understands this better than I do, but one clear difference is that the `numba` implementation is parallelized over the number of regular covariates.

**Comment by jeffgortmaker** [comment]
Ah okay, identical convergence criteria will help pinpoint the speedup. Algorithm too -- the new code is also non-accelerated MAP with Kaczmarz transform?

Parallelization is clearly not apples-to-apples. A good first benchmark would be comparing identical convergence criteria, identical algorithm, and in serial, not parallel. Should be possible to disable multithreading both in NumPy and Numba. (Checking memory usage will also be helpful since there's usually going to be a memory-speed tradeoff, but it's probably less relevant here.)

If it turns out most of the speedup (in a large problem, e.g. minutes of compute time) is from parallelization, then I think we should be having a different conversation about parallelization, not necessarily NumPy vs. Numba. 

Right now,  the only way I can think to use multiple cores with PyHDFE (beyond NumPy's "under the hood" multithreading) is spawning multiple Python processes (e.g. with the default multiprocessing module), passing an initialized `Algorithm` to each, and in each, calling `Algorithm.residualize` for a single column (or batch of columns) of the matrix to be residualized. The extra code required for this is fine in my mind, since PyHDFE isn't meant to be particularly user-friendly, but as an input into other packages, which could support or not support parallelization.

But if it turns out that people using PyHDFE really struggle with the additional overhead this creates, then Numba's parallelization could be a good option (are there others?), compared to Python's multiprocessing and (hopefully soon to be lessened) issues with the GIL. 

But of course this is all contingent on parallelization being a big part of the story here -- might not be!

**Comment by styfenschaer** [comment]
Hi @jeffgortmaker  and @s3alfisc 

I can't contribute much to this discussion since I'm not familiar with the real-world problems you're trying to solve with this package, but I can try. 

First of all, I just helped improve the performance of an interesting piece of code that was provided to me. It was a very quick design that can certainly be improved.

When it comes to performance _on a single machine,_ I can tell you that the speed increase comes mostly from better memory management and avoiding redundant work. There's nothing special about number crunching, nor have things like cache locality been optimized. In fact, the task is so simple that memory bandwidth is limited and using multiple threads provides a rather disappointing benefit. 

In terms of memory usage, this problem is fortunate because low memory usage goes hand in hand with high performance. Most of the memory required is the output matrix. The rest is vanishingly small for large problems. Theoretically, it would even be possible to perform the transform in place without requiring additional memory.

**Comment by jeffgortmaker** [comment]
Hi @styfenschaer, thanks so much for working on improving this! The main real-world problem is running regressions with lots of fixed effects, which gets a lot of use, particularly in the social sciences.

That's great news that most of the speed increase doesn't come from parallelization. Means that incorporating something like this into PyHDFE could really help out without requiring additional resources from users. Low memory usage is even better :)

@s3alfisc: In terms of next steps, I'm happy with the following:

1. Wait to see what the pure C++ implementation looks like.
2. Do some apples-to-apples comparisons (e.g., without parallelization) of it and a Numba approach.
3. Decide how to improve PyHDFE based on the results of this.

Looking forward!

**Comment by s3alfisc** [comment]
Sounds like a very good plan to me =)
